maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,204 @@
+    def __init__(
+        self, event_buffer: EventBuffer, topology: str, start_tick: int,
+        max_tick: int, snapshot_resolution: int, max_snapshots: int, additional_options: dict = {}
+    ):
+        super().__init__(
+            "data_center", event_buffer, topology, start_tick, max_tick,
+            snapshot_resolution, max_snapshots, additional_options
+        )
+
+        self._energy_consumption: int = 0
+        self._success_requirements: int = 0
+        self._total_latency: Latency = Latency()
+
+        self._init()
+
+    def _load_configs(self):
+        """Load configurations"""
+        with open(os.path.join(self._config_path, "config.yml")) as fp:
+            self._conf = safe_load(fp)
+
+    def _init(self):
+        # load config
+        self._load_configs()
+        self._register_events()
+        # PM init
+        self.machines: List[PhysicalMachine] = []
+        pm_amount = self._conf["pm_amount"]
+        cap_cpu = self._conf["pm_cap_cpu"]
+        cap_mem = self._conf["pm_cap_mem"]
+        self.machines = [PhysicalMachine(id=i, cap_cpu=cap_cpu, cap_mem=cap_mem) for i in range(pm_amount)]
+        self.vm: dict = {}
+
+    def step(self, tick: int):
+        """
+        Push business to next step.
+
+        Args:
+            tick (int): Current tick
+        """
+        # Load VM info into payload
+
+        # Update all live VMs memory utilization
+        self._update_vm_util(tick)
+        # Update all PM memory utilization
+        self._update_pm_util(tick)
+        # VM requirement event
+        vm_required_evt = self._event_buffer.gen_cascade_event(tick, Events.REQUIREMENTS, payload=None)
+        self._event_buffer.insert_event(vm_required_evt)
+
+    def get_metrics(self) -> dict:
+        """Get current enviornment metrics information.
+
+        Args:
+            dict: A dict that contains the environment metrics.
+
+        Returns:
+            dict: Metrics information.
+        """
+
+        energy_consumption = self._energy_consumption
+        success_requirements = self._success_requirements
+        total_latency = self._total_latency
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=energy_consumption,
+            success_requirements=success_requirements,
+            total_latency=total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(Events.REQUIREMENTS, self._on_vm_required)
+        self._event_buffer.register_event_handler(Events.FINISHED, self._on_vm_finished)
+
+        # Decision event
+        self._event_buffer.register_event_handler(MaroEvents.TAKE_ACTION, self._on_action_received)
+
+    def _update_vm_util(self, tick: int):
+        """Update all live VMs memory utilization."""
+        for _, vm in self.vm.items():
+            vm.util_mem = vm.show_util_series()[tick]
```

Comment:
if the vm launched at tick 10,

what will vm.show_util_series()[0] be?
vm.show_util_series()[10] will be the first util of this VM or which one?

Reply:
I think it is better to maintain the same util_series length in each VM, that is, if some VMs are created at 3th tick. Then it might be [-1, -1, -1, xxx, xxx, xxx, ...].

Reply:
but it will waste a lot of memory.

And if we support over-subscribe. If some VM is hang up when over-utilization happens, the series should be modified.

As we discussed before, how about using the be.tick - vm._start_tick as the index?
For the situation after an over-utilization happens, some more detailed design may be needed?